File: 0594. Longest Harmonious Subsequence/0594.java
// time Complexity: O(n)
// space Complexity: O(n)
public class Solution {
    public int findLHS(int[] nums) {
        // create a HashMap to count frequency of each number
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }

        int longest = 0;
        // iterate through keys and check if key + 1 exists
        for (int key : map.keySet()) {
            if (map.containsKey(key + 1)) {
                // update longest if current harmonious subsequence is longer
                longest = Math.max(longest, map.get(key) + map.get(key + 1));
            }
        }
        return longest;
    }
}

File: 0594. Longest Harmonious Subsequence/0594.py
# time complexity: O(nlogn) due to sorting
# space complexity: O(1)    
class Solution(object):
    def findLHS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # sort the array
        nums.sort()
        longest = 0
        i = 0  # left pointer
        
        # use the right pointer to scan through the sorted list
        for j in range(len(nums)):
            # move left pointer until the difference is not more than 1
            while nums[j] - nums[i] > 1:
                i += 1
            # only update if the current window is harmonious (diff == 1)
            if nums[j] - nums[i] == 1:
                longest = max(longest, j - i + 1)
        return longest